l2-contracts/contracts/AttesterRegistry.sol

```diff
@@ -0,0 +1,164 @@
+        // being able to be removed.
+        uint256 inactiveSince;
+    }
+
+    struct CommitteeAttester {
+        uint256 idx;
+        uint256 weight;
+        bytes pubKey;
+    }
+
+    modifier onlyOwner() {
+        require(msg.sender == owner, "Unauthorized: onlyOwner");
+        _;
+    }
+
+    constructor(address _owner) {
+        owner = _owner;
+    }
+
+    // Adds a new attester to the registry. Fails if an attester with the
+    // same public key already exists.
+    function add(
+        uint256 weight,
+        bytes calldata pubKey
+    ) external onlyOwner returns (uint256) {
+        // Check if validator with the same public key already exists.
+        for (uint256 i = 0; i < attesterIndexes.length; i++) {
+            uint256 idx = attesterIndexes[i];
+            require(!compareBytes(attesters[idx].pubKey, pubKey), "pubKey already exists");
+        }
+
+        attesterSequence++;
+        uint256 idx = attesterSequence;
+        attesters[idx] = Attester(weight, pubKey, false, 0);
+        attesterIndexes.push(idx);
+        return idx;
+    }
+
+    // Removes an attester. Should fail if the validator is still active or
+    // the inactivity delay has not passed.
+    function remove(uint256 idx) external onlyOwner {
+        verifyExists(idx);
+        require(attesters[idx].isInactive, "Attester is still active");
+        require(epoch >= attesters[idx].inactiveSince + INACTIVITY_DELAY,
+            "Attester's inactivity delay has not passed"
+        );
+
+        // Remove from mapping.
+        delete attesters[idx];
+
+        // Remove from array by swapping the last element (gas-efficient, not preserving order).
+        uint256 offset = idxOffset(idx);
+        attesterIndexes[offset] = attesterIndexes[attesterIndexes.length - 1];
+        attesterIndexes.pop();
+    }
+
+    // Inactivates an attester.
+    function inactivate(uint256 idx) external onlyOwner {
+        verifyExists(idx);
+        attesters[idx].isInactive = true;
+        attesters[idx].inactiveSince = epoch;
+    }
+
+    // Activates an attester.
+    function activate(uint256 idx) external onlyOwner {
+        verifyExists(idx);
+        attesters[idx].isInactive = false;
+    }
+
+    // Changes the weight.
+    function changeWeight(uint256 idx, uint256 weight) external onlyOwner {
+        verifyExists(idx);
+        attesters[idx].weight = weight;
+    }
+
+    // Changes the public key.
+    function changePublicKey(
+        uint256 idx,
+        bytes calldata pubKey
+    ) external onlyOwner {
+        verifyExists(idx);
+        attesters[idx].pubKey = pubKey;
+    }
+
+    // Creates a new committee list which will become the next committee list.
+    function setNextCommittee() external onlyOwner {
+        epoch += 1;
+
+        // Replace current committee with next committee.
+        delete committee;
+        for (uint256 i = 0; i < nextCommittee.length; i++) {
```

Comment:
Should there be some check that the next committee is not empty? I don't know if inactivating everyone to stop submissions is a use case.

Reply:
If all attesters inactivate themselves, there's really nothing we can do anyway.

Reply:
Okay, and I suppose the owner can reactivate them any time as the use of epochs for attesters is a bit unclear anyway. As long as there is no stretch where attestations _have_ to happen but nobody is legally able to send them any more, and reactivation won't help, we're good 👍